src/array.vala

```diff
@@ -2,70 +2,82 @@ using GLib;
+    private size_t* _shape_in;
     public size_t* shape {
         get {
             return _shape;
         }
         construct {
-            _shape = new size_t[dimension];
-            Memory.copy (_shape, value, dimension * sizeof (size_t));
+            _shape_in = value;
         }
     }
 
-    [CCode (array_length = false)]
-    ssize_t[] _strides;
-
+    /* bytes to skip for each dimensions, immutable. Initialized from constructor value later */
+    private ssize_t _strides[32];
+    private ssize_t * _strides_in;
     public ssize_t* strides {
         get {
             return _strides;
         }
         construct {
-            _strides = new ssize_t[dimension];
-            if (value == null) {
-                for (var i = dimension; i > 0; i--) {
-                    _strides[i - 1] = (i == dimension) ? (ssize_t) scalar_size : _strides[i] * (ssize_t) _shape[i];
-                }
-            } else {
-                Memory.copy (_strides, value, dimension * sizeof (ssize_t));
-            }
+            _strides_in = value;
         }
     }
 
-    public size_t origin { get; construct; default = 0; }
+    /* total number of scalar elements in the array, immutable */
+    public size_t size {get; private set;}
 
-    public size_t size {
-        get {
-            size_t size = 1;
-            for (var i = 0; i < dimension; i++) {
-                size *= _shape[i];
-            }
-            return size;
-        }
-    }
+    /* GObject that owns the memory buffer for storage of scalar elements */
+    public Bytes? data {get; construct; }
 
-    private          Bytes   _data;
-    internal unowned uint8* _cached_data;
+    /* pointer to the memory location of 0th element */
+    public uint8 * origin {get; construct; default = null;}
 
-    public Bytes? data {
-        get
-        {
-            return _data;
+
+    construct {
+        /* initializes the read-only attributes of the Array Object */
+        assert (_dimension <= 32);
+
+        /* We will copy the in-values from g_object_new */
+        for (var i = 0; i < _dimension; i ++) {
+            _shape[i] = _shape_in[i];
         }
-        construct {
-            _data        = value;
-            if (value != null) {
-                _cached_data = value.get_data ();
+
+        if (_strides_in == null) {
+            /* assume C contiguous strides */
+            for (var i = _dimension; i > 0; i--) {
+                _strides[i - 1] = (i == _dimension) ? (ssize_t) scalar_size : _strides[i] * (ssize_t) _shape[i];
+            }
+        } else {
+            for (var i = 0; i < _dimension; i ++) {
+                _strides[i] = _strides_in[i];
             }
         }
+
+        /* calculate size, since it is immutable, we do it once here */
+        _size = 1;
+        for (var i = 0; i < _dimension; i ++) {
+            _size *= _shape[i];
+        }
+
+        _data = _data ?? new Bytes (new uint8[scalar_size * _size]);
```

Comment:
This is going to break the current hack for PyGObject (a.k.a empty array). Maybe we could wait to get a proper fix for introspection and keep allocating in the `Array` constructor?

I also agree that once constructed, the data should be fully allocated.